### DISCUSSION TITLE
getting rsync'ed targets to not be "outdated"

### DISCUSSION THREAD

**Discussion by rmflight** [ropensci/targets#796] [Help] [answered]
Using {targets} to manage a workflow, which is great.

We don't have a proper cluster setup, but I have access to remote machines with much better specs than my laptop, so I can use git to keep the plan in sync locally and remotely.
I'm also using {renv} to manage the packages, etc, and those should be identical, except for the machine, and underlying versions of Linux.

When I want to work with something locally, I use `rsync` to move the files over.

```
rsync -avxP -e "ssh -p ..." remote:path/to/_targets .
```

When I query the remote cache with `tar_network`, I see that a bunch of my targets are "uptodate".

When I query the local cache after the `rsync` above, those same targets are "outdated".

I'm wondering if there is either better calls to `rsync` or certain arguments to `tar_network()`, or if this is a bug and the targets *should* stay as "uptodate" after an `rsync` like this?

**Comment by wlandau** [comment, marked as answer]
If you are syncing the code and the data store and packages and R versions and all external input/output files all together, then generally the targets will stay up to date on the new machine. If you are already doing that, then it there is probably something else specific to your project that is causing trouble. `tar_sitrep()` gives more specific information on why certain targets are not up to date.

By the way, instead of syncing back and forth, you could try the `clustermq` ssh connector: https://mschubert.github.io/clustermq/articles/userguide.html#ssh-connector. If `clustermq::Q()` on its own works with your setup, then you can set it up with `targets`: https://books.ropensci.org/targets/hpc.html#clustermq

**Reply by rmflight** [reply]
Thanks for the answer @wlandau. I looked at the output from `tar_sitrep()`, and didn't see anything particularly helpful, unfortunately.

More background: I want the workflow to depend on a package I've developed, and may need changes as I run targets and experience errors. Therefore, I'm trying to depend on the version number and githubsha1 from the `packageDescription`. So I have this target in my list:

```r
pc_info = utils::packageDescription("packageName")
pkg_tar = tar_target(pkg, c(pc_info$Version, pc_info$GithubSHA1))

data_tar = tar_target(data_1, run_data("path/to/datafile", pkg))

list(pkg_tar, data_tar)
```

Then all of the first layer of targets have `pkg` as one of their inputs into the function, so that they depend on it. This still seems to get *outdated* when I move them to another machine, unfortunately.

Regarding `clustermq`, our lab machines live behind a firewall machine w/ a custom ssh port, so it's local -> firewall-machine -> run-machine. From what I can tell from the `clustermq` docs, that's going to be a no go, unfortunately.

**Reply by wlandau** [reply]
Version numbers might have funny attributes, you could try `as.character(c(pc_info$Version, pc_info$GithubSHA1))` instead. Also, does anything depend on `pc_info`?

You might also consider https://books.ropensci.org/targets/packages.html#package-based-invalidation.

**Reply by rmflight** [reply]
Ah! The [package based invalidation](https://books.ropensci.org/targets/packages.html#package-based-invalidation) was exactly what I needed, and I also noticed some of my inputs used full paths instead of relative paths, which also created differences.

Thank you for all the pointers Will!
